[ENT-640] Ensure all Enterprise/Consent API endpoints apply appropriate permissioning/filtering scheme

UmanShahzad opened the pull request:
**Description:** This PR adds proper permission/filter schemes to our API endpoints.

**JIRA:** [ENT-640](https://openedx.atlassian.net/browse/ENT-640)

**Merge deadline:** ASAP

**Testing instructions:**

Please see the thread on [the ticket](https://openedx.atlassian.net/browse/ENT-640) to understand how our endpoints should actually be in fact behaving. Then, go through each of the APIs as a staff & non-staff user, and at each step ensuring you get the expected results.

Note that we will **skip** the consent API for testing here -- no changes were made to it that should affect its current functionality in any way, except making sure it works in cases where it may have normally broke (i.e. getting back-up values).

As a **staff user**, go through the following endpoints and ensure you get back all objects for that endpoint:

1. `/enterprise/api/v1/enterprise-catalogs/`
1. `/enterprise/api/v1/enterprise-course-enrollment/`
1. `/enterprise/api/v1/enterprise-customer/`
1. `/enterprise/api/v1/enterprise-customer/{uuid}/courses/`
1. `/enterprise/api/v1/enterprise-customer-branding/`
1. `/enterprise/api/v1/enterprise-customer-entitlement/`
1. `/enterprise/api/v1/enterprise-learner/`
1. `/enterprise/api/v1/catalogs/`

Now make a fresh **non-staff** user (confirm activation email) who isn't connected to any Enterprise, and get any Enterprise UUID for testing. Go through the following endpoints:

1. `/enterprise/api/v1/enterprise-catalogs/` -- ensure that you see **no catalogs**.
1. `/enterprise/api/v1/enterprise-course-enrollment/` -- ensure that you see **no enrollments**.
1. `/enterprise/api/v1/enterprise-customer/` -- ensure that you see **no customers**.
1. `/enterprise/api/v1/enterprise-customer/{uuid}/courses/` -- ensure that you see **no customer courses**.
1. `/enterprise/api/v1/enterprise-customer-branding/` -- ensure that you see **no brandings**.
1. `/enterprise/api/v1/enterprise-customer-entitlement/` -- ensure that you see **no entitlements**.
1. `/enterprise/api/v1/enterprise-learner/` -- ensure that you see **no learners**.
1. `/enterprise/api/v1/catalogs/` -- ensure that you see **no catalogs**.

Now make your **non-staff** user become connected to some Enterprise. Get the UUID for testing. Go through the following endpoints:

1. `/enterprise/api/v1/enterprise-catalogs/` -- ensure that you see only the **catalogs of the customer you're associated with**.
1. `/enterprise/api/v1/enterprise-course-enrollment/` -- ensure that you see **no enrollments** (because you've not enrolled in anything yet).
1. `/enterprise/api/v1/enterprise-customer/` -- ensure that you only see the **customer you're associated with**.
1. `/enterprise/api/v1/enterprise-customer/{uuid}/courses/` -- ensure that you see **no courses** (unable to fetch them because your user isn't given permissions for discovery).
1. `/enterprise/api/v1/enterprise-customer-branding/` -- ensure that you only see the **branding of the customer you're associated with**.
1. `/enterprise/api/v1/enterprise-customer-entitlement/` -- ensure that you only see the **entitlements of the customer you're associated with**.
1. `/enterprise/api/v1/enterprise-learner/` -- ensure that you only see **yourself**.
1. `/enterprise/api/v1/catalogs/` -- ensure that you see **no catalogs** (unable to fetch them because your user isn't given permissions for discovery).

Now go to the course enrollment landing page for any course for your Enterprise, and enroll in it with your **non-staff** user.

1. `/enterprise/api/v1/enterprise-course-enrollment/` -- ensure that you see **that enrollment** only.

**Merge checklist:**

- [ ] Check that the versions of the requirements in the `platform-master.in` file match edx-platform.
- [ ] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [ ] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are (reasonably) squashed
- [ ] Translations are updated
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

edx-webhook commented:
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-enterprise&number=193"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

douglashall commented:
@UmanShahzad Would you mind creating a separate PR with just the removal of the auth-user endpoint? I would like to land that one ASAP without worrying about reviewing the other changes.

UmanShahzad commented:
@douglashall Yes, that sounds like a plan. Should I include removing the `site` endpoint in it?

douglashall commented:
Let's just stick to the auth-user for now. Thanks.

UmanShahzad commented:
@douglashall @brittneyexline I've done all the main code changes in this now, but tests still need to be updated. Can you guys please take a look before I do that, to make sure the new permissions/filters are headed in the right direction?

UmanShahzad commented:
Rebased on `master` from 8714b2fbfc3eb7fe2811e0fd5f60757ec5b773e4

UmanShahzad commented on enterprise/api/v1/urls.py:
```diff
@@ -9,7 +9,7 @@
 
 router = DefaultRouter()  # pylint: disable=invalid-name
-router.register("site", views.SiteViewSet, 'site')
+router.register("auth-user", views.UserViewSet, 'auth-user')
```
Ok.. rebase's auto-merge didn't work as expected.

UmanShahzad commented:
Squashed from 9c26160a7ffa686f245951c9b25ace3159b1a7fb

douglashall commented on enterprise/api/permissions.py:
```diff
@@ -14,7 +14,7 @@
 
 
-class IsServiceUserOrReadOnly(permissions.IsAuthenticated):
+class IsServiceUserOrReadOnly(permissions.DjangoModelPermissions):
```
Actually, I don't think we need this class at all if were are using `DjangoModelPermissions`. We will just need to make sure that the service user in sandbox/stage/prod has the appropriate permissions.

UmanShahzad replied:
Ah, this is correct. As per your comment below, we can just include it in the `permissions_classes`.

douglashall commented on enterprise/api/v1/views.py:
```diff
@@ -63,6 +62,7 @@ class EnterpriseModelViewSet(EnterpriseViewSet):
     """
 
     filter_backends = (filters.OrderingFilter, filters.DjangoFilterBackend)
```
We should add `permissions_classes=(DjangoModelPermissions, )` here.

UmanShahzad replied:
:+1:

UmanShahzad replied:
Actually this would get overwritten below, so placing it in both `EnterpriseReadOnlyModelViewSet` and `EnterpriseReadWriteModelViewSet`.

UmanShahzad commented on enterprise/api/v1/views.py:
```diff
@@ -87,6 +87,9 @@ class EnterpriseCustomerViewSet(EnterpriseReadOnlyModelViewSet):
     serializer_class = serializers.EnterpriseCustomerSerializer
+    permission_classes = (permissions.IsAuthenticated, IsServiceUserOrReadOnly, IsStaffUserOrLinkedToEnterprise,)
+    filter_backends = (filters.OrderingFilter, filters.DjangoFilterBackend, UserFilterBackend)
+    ID_FILTER = 'enterprise_customer_users__user_id'
```
Yes that's a better name, thanks!

UmanShahzad commented:
@douglashall Can you give this another round of code review? Honestly there seemed to be a lot of repetitive or unnecessary stuff in our code that was really hard for me to track down, and that's probably why we totally missed that filtering/permission schemes were not so good in the first place.

UmanShahzad commented on tests/test_enterprise/api/test_serializers.py:
```diff
@@ -395,3 +397,37 @@ def test_update_enterprise_courses(self, mock_config_helpers, _):
+        Success for POSTing with users (determined by username) depends on whether the user exists.
+        """
+        self.user.is_staff = True
+        self.user.save()
```
Note to self: move this out into `setUp`.

itsjeyd commented on tests/test_enterprise/api/test_views.py:
```diff
@@ -629,28 +688,49 @@ def test_enterprise_customer_catalogs_list(self, is_staff, is_linked_to_enterpri
+            True,
             fake_enterprise_api.build_fake_enterprise_catalog_detail(include_enterprise_context=True),
         ),
     )
```
@UmanShahzad There's a few combinations missing here:

* `is_staff`: `False`, `is_linked_to_enterprise`: `True`, `has_permissions`: `False`
* `is_staff`: `True`, `is_linked_to_enterprise`: `True`, `has_permissions`: `False`
* `is_staff`: `True`, `is_linked_to_enterprise`: `False`, `has_permissions`: `False`

Would any of these make sense to test, in addition to the cases that are already covered?

UmanShahzad replied:
Sure, I'll add them and take a look! :+1:

UmanShahzad replied:
So yeah, as below, turned out this was obsolete :P

UmanShahzad commented on consent/api/permissions.py:
```diff
@@ -14,10 +14,16 @@ class IsUserInRequest(permissions.BasePermission):
+            username_in_request = request.data.get(
+                'username',
+                request.query_params.get('username', '')
+            )
```
This is more about being cautious and imitating the way we extract certain parameters in [`get_required_query_params`](https://github.com/edx/edx-enterprise/blob/master/consent/api/v1/views.py#L95). It may be useful to write a function to reuse this type of extraction, so... *note to self: write a function for this*.

itsjeyd replied:
Sounds good! That function should probably use [`QUERY_PARAM_METHODS`](https://github.com/edx/edx-enterprise/blob/master/consent/api/v1/views.py#L81) (which would need to be available globally/outside the definition of `DataSharingConsentView`)  instead of explicitly listing `GET` and `DELETE`.

UmanShahzad replied:
Actually not sure if the function itself should be using that, because what I had in mind of abstracting with this function was the actual body of those `if`/`else` statements. I'm not sure if we'd use that function here if it was extracting things like `program_uuid` and stuff. So instead I'm simply thinking of reusing this:

```python
def get_request_value(request, value, default=None):
    return request.query_params.get(
        value,
        request.data.get(value, default)
    )
```

UmanShahzad replied:
Ok nevermind, yeah we can use `QUERY_PARAM_METHODS`, sorry :P

itsjeyd replied:
@UmanShahzad If `QUERY_PARAM_METHODS` is defined as `{'GET', 'DELETE'}` (and is available globally/outside the definition of `DataSharingConsentView`), we could do the following, right?

```python
from ... import QUERY_PARAM_METHODS

# ...

def get_request_value(request, key, default=None):
    """
    Get the value in the request, either through query parameters or posted data, from a key.

    :param request: The request from which the value should be gotten from.
    :param key: The key to use to get the desired value.
    :param default: The backup value to use in case the input key cannot help us get the value.
    :return: The value we're looking for.
    """
    if request.method in QUERY_PARAM_METHODS:
        return request.query_params.get(key, request.data.get(key, default))
    else:
        return request.data.get(key, request.query_params.get(key, default))
```

That's what I was getting at :) Sorry if that wasn't clear!

UmanShahzad replied:
Ah I just missed this comment while addressing the review.

Yeah we can do this, but I'm not sure what would be a reasonable place for `QUERY_PARAM_METHODS` to live, that's all. Maybe in `enterprise.constants`? Either way this is not that big of a deal because it's not being reused right now. If you're okay with it, we can come back to this later once it becomes necessary to reuse.

itsjeyd replied:
Fine with me :)

UmanShahzad commented on enterprise/api/filters.py:
```diff
@@ -9,40 +9,49 @@
+class UserFilterBackend(filters.BaseFilterBackend):
     """
-    Allow filtering based on user's email or username on enterprise customer user api endpoint.
+    Filter backend for any view that needs to filter against the request user's ID.
```
Yes, sure! :+1:

UmanShahzad commented on enterprise/api/v1/views.py:
```diff
@@ -77,7 +78,7 @@ class EnterpriseReadWriteModelViewSet(EnterpriseModelViewSet, viewsets.ModelView
     """
 
-    permission_classes = (permissions.IsAuthenticated, IsServiceUserOrReadOnly,)
+    permission_classes = (permissions.IsAuthenticated, permissions.DjangoModelPermissions, IsServiceUserOrReadOnly,)
```
Yeah, it's basically because these are *model* permissions and not every view set inheriting from `EnterpriseViewSet` is necessarily going to be a model's view set (we have `EnterpriseModelViewSet` for that).

itsjeyd replied:
Ah, OK. Thanks for the details!

douglashall replied:
Didn't we decide we no longer need `IsServiceUserOrReadOnly` now that we are using `DjangoModelPermissions`?

UmanShahzad replied:
@douglashall Yep sorry, you're right about that.

UmanShahzad commented on enterprise/api/v1/views.py:
```diff
@@ -88,27 +89,23 @@ class EnterpriseCustomerViewSet(EnterpriseReadOnlyModelViewSet):
-    @detail_route(
-        permission_classes=[permissions.IsAuthenticated, IsServiceUserOrReadOnly, IsStaffUserOrLinkedToEnterprise],
-        authentication_classes=[JwtAuthentication, BearerAuthentication, SessionAuthentication],
-        throttle_classes=[ServiceUserThrottle],
```
As far as I can tell, these were not necessary in the first place (although it could've been that these *were* necessary at first but code moved around and someone didn't realize to erase this); they already come just fine through our view set inheritance hierarchy (from `EnterpriseViewSet`).

itsjeyd replied:
Sounds good, thanks!

UmanShahzad commented on tests/test_enterprise/api/test_filters.py:
```diff
@@ -13,30 +13,63 @@
+        ('fake-uuid', {'detail': 'Not found.'})
+    )
+    @ddt.unpack
+    def test_filter_for_detail(self, uuid, expected_content_in_response):
```
Ah I forgot the docstring here, thanks! :+1: 

I think you bring up a good point about distinguishing between staff/non-staff here. We should probably cover this based off of staff status rather than a valid UUID because we don't care about UUID validity here, we're just filtering.

UmanShahzad commented on tests/test_enterprise/api/test_filters.py:
```diff
@@ -45,22 +78,31 @@ def setUp(self):
+        ('', TEST_EMAIL, True, 1),
+        ('', '', True, 1),
+        ('dummy', '', True, 0),
+        ('dummy', 'dummy@example.com', True, 0),
```
Sure!

UmanShahzad commented on tests/test_enterprise/api/test_filters.py:
```diff
@@ -45,22 +78,31 @@ def setUp(self):
+        ('', TEST_EMAIL, False, 1),
+        ('', '', False, 1),
+        ('dummy', '', False, 1),
+        ('dummy', 'dummy@example.com', False, 1),
```
:+1:

UmanShahzad commented on tests/test_enterprise/api/test_serializers.py:
```diff
@@ -395,3 +397,37 @@ def test_update_enterprise_courses(self, mock_config_helpers, _):
+        self.user.is_staff = True
+        self.user.save()
+        permission = Permission.objects.get(name='Can add Enterprise Customer Learner')
+        self.user.user_permissions.add(permission)
```
Yeah this is handled by the `DjangoModelPermissions` that we added, rather than by us. It will check for permissions for certain unsafe methods like `POST`, `DELETE`, etc.. This is something @douglashall recommended and I really like the idea.

UmanShahzad commented on tests/test_enterprise/api/test_views.py:
```diff
@@ -221,7 +242,30 @@ def test_enterprise_learner_entitlements(
+            403
+        ),
+        (
             # Test a bad request due to no existing EnterpriseCustomerUser
```
:P I like drive-by fixes

UmanShahzad commented on tests/test_enterprise/api/test_views.py:
```diff
@@ -296,33 +342,28 @@ def test_get_enterprise_customer_user_contains_consent_records(self):
+            permission = Permission.objects.get(name='Can add Enterprise Customer Learner')
+            self.user.user_permissions.add(permission)
 
+        response = self.client.post(settings.TEST_SERVER + reverse('enterprise-learner-list'), data=data)
```
Woops, yeah, will take care of that.

UmanShahzad commented on tests/test_enterprise/api/test_views.py:
```diff
@@ -360,16 +411,18 @@ def test_post_enterprise_customer_user_logged_out(self):
             False,
             {},
-            {'detail': 'User must be a staff user or associated with the specified Enterprise.'}
+            {'detail': 'User {} is not associated with an EnterpriseCustomer.'.format(TEST_USERNAME)}
```
So I think I made a slight mistake here after reading this. The first boolean corresponds to staff status, which should be enough to let us through to see an Enterprise Customer's courses (AFAIK). This is due to a change I made that I thought would be harmless but actually not. So I'll make that change and see that the error messages make more sense.

UmanShahzad commented on tests/test_enterprise/api/test_views.py:
```diff
@@ -382,6 +435,16 @@ def test_post_enterprise_customer_user_logged_out(self):
             True,
+            False,
+            {},
+            {'detail': 'User {} is not associated with an EnterpriseCustomer.'.format(TEST_USERNAME)},
```
Yep, see above. I have to fix something in the logic (not much, just 1-2 lines).

UmanShahzad commented on tests/test_enterprise/api/test_views.py:
```diff
@@ -667,14 +747,17 @@ def test_enterprise_customer_catalogs_detail(self, is_staff, is_linked_to_enterp
                 enterprise_customer=enterprise_customer
             )
+        if has_permissions:
+            permission = Permission.objects.get(name='Can add Enterprise Customer Catalog')
```
I actually think this may now be obsolete. I had this earlier because I thought it was necessary, but after taking another look, I don't think so.

UmanShahzad commented:
@itsjeyd I've addressed your review. I'll add testing instructions and post when I'm done with that in a new comment.

AFAIK, yeah, this behavior (I hope I implemented it correctly?) is something @douglashall has agreed to (especially the `DjangoModelPermissions` part, which I think is great). But obviously with the code changes that have occurred, I'd like to double check with him here when he gives this another round of review.

UmanShahzad commented:
This is now on the sandbox, but I have yet to update the test instructions.

cc @douglashall in case there are any permissions we might have to add to our service users on the sandbox now.

itsjeyd commented on tests/test_enterprise/api/test_views.py:
```diff
@@ -381,6 +448,7 @@ def test_post_enterprise_customer_user_logged_out(self):
             1,
             ENTERPRISE_CUSTOMER_COURSES_ENDPOINT,
+            False,
             True,
```
@UmanShahzad The updates to the data for this test look great. I think we should add one more case: `is_staff`: `True`, `is_linked_to_enterprise`: `False`, `mocked_catalog_courses`: `fake_catalog_api.FAKE_CATALOG_COURSE_PAGINATED_RESPONSE`.

UmanShahzad replied:
:+1:

UmanShahzad commented on consent/helpers.py:
```diff
@@ -99,3 +99,17 @@ def get_program_data_sharing_consent(username, program_uuid, enterprise_customer
     return ProxyDataSharingConsent.from_children(program_uuid, *child_consents)
+
+
+def get_request_value(request, key, default=None):
```
Sure. :+1:

UmanShahzad commented on consent/helpers.py:
```diff
@@ -99,3 +99,17 @@ def get_program_data_sharing_consent(username, program_uuid, enterprise_customer
+    """
+    Get the value in the request, either through query parameters or posted data, from a key.
+
+    :param request: The request from which the value should be gotten from.
```
:+1:

UmanShahzad commented on tests/test_consent/test_helpers.py:
```diff
@@ -26,3 +27,18 @@ def test_get_data_sharing_consent_no_enterprise(self):
+        Request value is retrieved from the posted data and not from the query parameters for certain methods.
+        """
+        request = mock.MagicMock(method='POST', query_params={'key': 'query_params'}, data={'key': 'data'})
+        assert helpers.get_request_value(request, 'key') == 'data'
```
:+1:

UmanShahzad commented on tests/test_enterprise/api/test_filters.py:
```diff
@@ -13,30 +13,76 @@
+        self.user.is_staff = is_staff
+        self.user.save()
+        factories.EnterpriseCustomerFactory()
+        response = self.client.get(settings.TEST_SERVER + reverse('enterprise-customer-list'))
```
Hmm, not really, no. This just unit tests the filter, which has the same set of logic for all endpoints, and that's what we're really testing, so what endpoint is used is irrelevant as long as that endpoint uses this filter.

itsjeyd replied:
@UmanShahzad OK, let's add a note about that to the docstring:

```
Filter objects based off of user ID -- show all for staff, but only the authenticated user himself otherwise.

Note that the choice of endpoint below is arbitrary.
This test could use any endpoint that makes use of the ``UserFilterBackend`` filter.
```

UmanShahzad replied:
:+1:

UmanShahzad commented on tests/test_enterprise/api/test_filters.py:
```diff
@@ -13,30 +13,76 @@
+        if not is_linked:
+            self.enterprise_customer_user.user_id = self.user.id + 1
+            self.enterprise_customer_user.save()
+        response = self.client.get(settings.TEST_SERVER + reverse('enterprise-customer-detail', (FAKE_UUIDS[0],)))
```
Nope, not as far as testing the filter is concerned -- which endpoint to choose is quite arbitrary.

itsjeyd replied:
@UmanShahzad OK, let's add a note about that to the docstring:

```
Filter the specific object based off of user ID.

Show the object for staff, and for the authenticated user if there's an association with the object
and the user through the view's `USER_ID_FILTER`.

Note that the choice of endpoint below is arbitrary.
This test could use any endpoint that makes use of the ``UserFilterBackend`` filter.
```

UmanShahzad replied:
:+1:

UmanShahzad commented:
@itsjeyd @douglashall Thanks for the review guys! I've addressed it and pushed. I'm going to rebase and then post the testing instructions (finally...).

UmanShahzad commented:
Rebased from 7d1cd2816f4819d84dc4fb6b213c1b8ee802021e on `master`.

UmanShahzad commented:
@itsjeyd @douglashall Testing instructions are up. It's possible that I missed something because there are so many different endpoints. Do let me know!

UmanShahzad commented:
@douglashall I think we have an issue where if the user is not allowed to see enterprise customer data if not already linked to the customer, then they can't go through SSO registration through that customer because https://github.com/edx/edx-platform/blob/master/openedx/features/enterprise_support/api.py#L365 throws the unauthorized error.

UmanShahzad commented on consent/api/permissions.py:
```diff
@@ -14,17 +15,7 @@ class IsUserInRequest(permissions.BasePermission):
-                request.query_params.get('username', '')
-            )
-        if request.user.username == username_in_request:
+        if request.user.username == get_request_value(request, 'username', ''):
```
Argh, totally missed this one during addressing review, too. So I can do this, but I feel a little weird about it: it's almost idiomatic that in `get_*` functions there is some default fallback and it's usually given as the last argument before explicit keyword args, or it's ignored for the default of the default (usually `None`).

UmanShahzad commented:
@douglashall The simplest, temporary solution I see to the above is to get the customer from the DB rather than the endpoint.

douglashall commented:
@UmanShahzad Or we should be using the service user to make the API call.

UmanShahzad commented:
@douglashall Yeah, or that :P

UmanShahzad commented:
So yeah, nevermind the above chat. My intuition failed me here but it def. seemed related because the issue was permissions related, which is what this PR touches.

Let's see if we run into any kind of related issue again after the revert is on the sandbox.

UmanShahzad commented:
@itsjeyd @douglashall I went through all of the testing instructions and things work for me. Let me know if you need any help with configuration setup etc..

P.S. I added permissions to our worker/service users on the LMS.

UmanShahzad commented:
@itsjeyd The `/enterprise/api/v1/enterprise-learner/` endpoint's serializer uses the customer serializer for the customer object, which, yeah, gives back all learners for that customer. I didn't originally design this serializer so I don't know what's the plan with that -- @douglashall what do you think?

The reason for `/enterprise/api/v1/enterprise-customer/{uuid}/courses/` returning 'Not found' is because it looks for an Enterprise Customer association first, and if it can't find the association it'll say "Not found", referring to the customer it was looking for. The other endpoints are mostly model endpoints which do filtering, so they return a count of 0 in that case. This is a small detail that I don't think is that important to deal with, though.

itsjeyd commented:
Thanks for the details, @UmanShahzad!

> The reason for /enterprise/api/v1/enterprise-customer/{uuid}/courses/ returning 'Not found' is because it looks for an Enterprise Customer association first, and if it can't find the association it'll say "Not found", referring to the customer it was looking for. The other endpoints are mostly model endpoints which do filtering, so they return a count of 0 in that case. This is a small detail that I don't think is that important to deal with, though.

Agreed :)

@douglashall If you could chime in on the remaining question about behavior of the `/enterprise/api/v1/enterprise-learner/`, that would allow us to move forward with this. Thanks!

douglashall commented:
@itsjeyd Good catch!

@UmanShahzad I think what we want to do is use a different EnterpriseCustomer serializer with `EnterpriseCustomerUserReadOnlySerializer` and `EnterpriseCustomerUserEntitlementSerializer`. This serializer would not include the `enteprise_customer_users` field (and probably some others as well). You could call it `MinimalEnterpriseCustomer` to follow the pattern we have in the discovery service.

It might make sense to merge this now and follow up with this work later, since we aren't making anything worse here with the current changes. Could you file a followup ticket if you decide to go that route?

UmanShahzad commented:
@douglashall Yeah I think that's a nice convention to follow, to use `MinimalEnterpriseCustomer`. However I'm not sure if that helps in the end of the day: although it would remove the full list of users for the `enterprise-learner` endpoint, it would still exist on the `enterprise-customer` endpoint, which the user could just as well query after getting the UUID through `enterprise-learner`, is that right?

UmanShahzad commented:
One alternative is to use different serializers depending on staff status. Non-staff users should always be given the `MinimalEnterpriseCustomer` serializer.

douglashall commented:
@UmanShahzad Good point. Perhaps your original suggestion is best. Remove the `enterprise_customer_users` field entirely (and maybe some others as well) from `EnterpriseCustomerSerializer`. We'll need to make sure we aren't using that data anywhere today though. Again, sounds like a followup ticket unless it is quick to check.

douglashall commented:
@UmanShahzad Yes, I think either solution would work. I think if we aren't making use of `enterprise_customer_users` currently, we just remove it. If we need to add it back in later on with appropriate filtering based on user permission we can without too much pain.

UmanShahzad commented:
@douglashall I'm totally in favor of just removing it and adding it back when it becomes necessary for some future functionality. I looked through `edx-platform` and `ecommerce`, and it doesn't look like we're using it. It's just placed in docstrings and mocks as what is expected out of the API, but not actually used anywhere.

So based off that, would you like me to rebase, erase `enterprise_customer_users` from `edx-enterprise`'s serializers, and call this good to merge without further testing? If we find use for this in the future we can file a ticket to not only add it back but do so in a nicer manner, i.e. with `MinimalEnterpriseCustomer`.

From this, I see we have a *lot* of improvement we can do to our endpoints. As we grow this codebase and look to become a service, this will probably become a higher and higher priority!

douglashall commented:
@UmanShahzad Agreed all around. Yes, let's go ahead and remove `enterprise_customer_users`. As always, thanks for digging into this.

UmanShahzad commented:
@douglashall Awesome! And no problem!

UmanShahzad commented:
Rebased from 4bbcad84d2684032a6b55d48e951e8e34c4e04a7 on `master`.

UmanShahzad commented:
Squashed from ffacbcf74afb4eee1b01f271e99d538e938f8081.

UmanShahzad commented:
I'm going to wait on version bumping & merging this until we get the safe deployment of the account-level consent code deletion out of the way. That way we can handle this in lock-step rather than have to forcefully merge this in with the migrations.

UmanShahzad commented:
@itsjeyd Thanks a lot, especially for the ecommerce PR! That's helpful.

And yep, when we finish safely deploying the migrations to delete account-level consent fields/models, we can then safely merge this and the version bump for edx-platform will include those removals of `enterprise_customer_users` from docstrings/mocks.

UmanShahzad commented:
I will be merging this soon today. But still needs a changelog etc..

UmanShahzad commented:
Rebased on `master` from 6e66c7398d5e940c7124043f7107c44150512790.
